README
Ticketmaster-API

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it to get an idea of the module's uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Ticketmaster::API

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Ticketmaster-API

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Ticketmaster-API

    CPAN Ratings
        http://cpanratings.perl.org/d/Ticketmaster-API

    Search CPAN
        http://search.cpan.org/dist/Ticketmaster-API/


LICENSE AND COPYRIGHT

Copyright (C) 2016 Erik Tank

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Changes
Revision history for Ticketmaster-API

0.01    Date/time
        First version, released on an unsuspecting world.



t/recordings/record.pl
#!perl

use 5.006;

use strict;
use warnings;
use lib '../../lib/';

use Storable;

use Ticketmaster::API;

my $real_api_key  = 'realAPIkey';
my $save_api_key  = 'testAPIkey';
my $mode = 'GET';
my $path_template = 'discovery/%s/venues/341396.json';
my %parameters = (
);

my $recording_name = 'GET_' . sprintf($path_template, 'v1') . "?apikey=$save_api_key";
foreach my $key (sort {$a cmp $b} keys %parameters) {
    $recording_name .= '_' . $key . '_' . $parameters{$key};
}
$recording_name =~ s/[=?&\/]/_/g;

die("record.pl needs to be run within the recordings directory") unless $0 eq 'record.pl';
die("recording already exists: $recording_name\n") if -e $recording_name;

my $obj = Ticketmaster::API->new(api_key => $real_api_key);
my $res = $obj->get_data(mode => $mode, path_template => $path_template);

store $res, $recording_name;

print "$recording_name populated\n";


t/Mock/Ticketmaster/API.pm
package Mock::Ticketmaster::API;

use strict;
use warnings;

use Storable;
use Test::Mock::Simple;

my $mock = Test::Mock::Simple->new( module => 'Ticketmaster::API', allow_new_methods => 1);

# Requires: mode, path_template (sprintf string), parameters (hash ref)
$mock->add( get_data => sub {
    my $self = shift;
    my %args = @_;

    my $mode = $args{mode} || Carp::croak("No mode provided (GET)");
    my $path_template = $args{path_template} || Carp::croak("No URI template provided");
    my %parameters = exists $args{parameters} ? %{$args{parameters}} : ();

    my $uri .= sprintf($path_template, $self->version());

    $uri .= '?apikey=' . $self->api_key();

    foreach my $key (sort {$a cmp $b} keys %parameters) {
        $uri .= '&' . $key . '=' . $parameters{$key};
    }

    $uri =~ s/[?&\/=]/_/g;

    my $mock_file_name = 't/recordings/' . $mode . '_' . $uri;

    my $ret = retrieve($mock_file_name);

    return $ret;
});

1;


lib/Ticketmaster/API/Discovery.pm
package Ticketmaster::API::Discovery;

use 5.006;
use strict;
use warnings;

=head1 NAME

Ticketmaster::API::Discovery - The great new Ticketmaster::API::Discovery!

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';
use parent 'Ticketmaster::API';

my $base_uri = 'discovery/%s';

=head1 SYNOPSIS

Quick summary of what the module does.

Perhaps a little code snippet.

    use Ticketmaster::API::Discovery;

    my $foo = Ticketmaster::API::Discovery->new();
    ...

=head1 EXPORT

A list of functions that can be exported.  You can delete this section
if you don't export anything, such as for a purely object-oriented module.

=head1 SUBROUTINES/METHODS

=head2 search_events

Also See: http://ticketmaster-api.github.io/products-and-docs/apis/discovery/#srch-events

=cut

sub search_events {
    my $self = shift;
    my $args = {@_};

    return $self->get_data(mode => 'GET', path_template => "$base_uri/events.json", parameters => $args);
}

#http://ticketmaster-api.github.io/products-and-docs/apis/discovery/#event-details
sub event_details {
    my $self = shift;
    my $args = {@_};

    my $id = delete $args->{id} || die("No event id provided\n");

    return $self->get_data(mode => 'GET', path_template => "$base_uri/events/$id.json", parameters => $args);
}

#http://ticketmaster-api.github.io/products-and-docs/apis/discovery/#event-img
sub event_images {
    my $self = shift;
    my $args = {@_};

    my $id = delete $args->{id} || die("No event id provided\n");

    return $self->get_data(mode => 'GET', path_template => "$base_uri/events/$id/images.json", parameters => $args);
}

#http://ticketmaster-api.github.io/products-and-docs/apis/discovery/#search-attractions
sub search_attractions {
    my $self = shift;
    my $args = {@_};

    return $self->get_data(mode => 'GET', path_template => "$base_uri/attractions.json", parameters => $args);
}

#http://ticketmaster-api.github.io/products-and-docs/apis/discovery/#attraction-details
sub attraction_details {
    my $self = shift;
    my $args = {@_};

    my $id = delete $args->{id} || die("No attraction id provided\n");

    return $self->get_data(mode => 'GET', path_template => "$base_uri/attractions/$id.json", parameters => $args);
}

#http://ticketmaster-api.github.io/products-and-docs/apis/discovery/#search-categories
sub search_categories {
    my $self = shift;
    my $args = {@_};

    return $self->get_data(mode => 'GET', path_template => "$base_uri/categories.json", parameters => $args);
}

#http://ticketmaster-api.github.io/products-and-docs/apis/discovery/#category-details
sub category_details {
    my $self = shift;
    my $args = {@_};

    my $id = delete $args->{id} || die("No category id provided\n");

    return $self->get_data(mode => 'GET', path_template => "$base_uri/categories/$id.json", parameters => $args);
}

#http://ticketmaster-api.github.io/products-and-docs/apis/discovery/#search-venues
sub search_venues {
    my $self = shift;
    my $args = {@_};

    return $self->get_data(mode => 'GET', path_template => "$base_uri/venues.json", parameters => $args);
}

#http://ticketmaster-api.github.io/products-and-docs/apis/discovery/#venue-details
sub venue_details {
    my $self = shift;
    my $args = {@_};

    my $id = delete $args->{id} || die("No venue id provided\n");

    return $self->get_data(mode => 'GET', path_template => "$base_uri/venues/$id.json", parameters => $args);
}

=head1 AUTHOR

Erik Tank, C<< <erik.tank at ticketmaster.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-ticketmaster-api-discovery at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Ticketmaster-API-Discovery>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Ticketmaster::API::Discovery


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Ticketmaster-API-Discovery>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Ticketmaster-API-Discovery>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Ticketmaster-API-Discovery>

=item * Search CPAN

L<http://search.cpan.org/dist/Ticketmaster-API-Discovery/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2016 Erik Tank.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


=cut

1; # End of Ticketmaster::API::Discovery


lib/Ticketmaster/API.pm
package Ticketmaster::API;

use 5.006;
use strict;
use warnings;

use Carp;
use JSON::XS;
use LWP::UserAgent;

=head1 NAME

Ticketmaster::API - The great new Ticketmaster::API!

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';


=head1 SYNOPSIS

Quick summary of what the module does.

Perhaps a little code snippet.

    use Ticketmaster::API;

    my $foo = Ticketmaster::API->new();
    ...

=head1 EXPORT

A list of functions that can be exported.  You can delete this section
if you don't export anything, such as for a purely object-oriented module.

=head1 SUBROUTINES/METHODS

=head2 function1

=cut

sub new {
    my $self;
    my $package = shift;
    my $class = ref($package) || $package;

    $self = {@_};
    bless($self, $class);

    $self->base_uri('https://app.ticketmaster.com') unless $self->base_uri;
    $self->version('v1') unless $self->version;
    
    Carp::croak("No api_key provided") unless $self->api_key();

    return $self;
}

sub base_uri {
    my $self = shift;

    $self->{base_uri} = shift if @_;

    return $self->{base_uri};
}

sub version {
    my $self = shift;

    $self->{version} = shift if @_;

    return $self->{version};
}

sub api_key {
    my $self = shift;

    $self->{api_key} = shift if @_;

    return $self->{api_key};
}

# Requires: mode, path_template (sprintf string), parameters (hash ref)
sub get_data {
    my $self = shift;
    my %args = @_;

    my $mode = $args{mode} || Carp::croak("No mode provided (GET)");
    my $path_template = $args{path_template} || Carp::croak("No URI template provided");
    my %parameters = exists $args{parameters} ? %{$args{parameters}} : ();

    my $uri = $self->base_uri;
    $uri .= '/' unless $uri =~ /\/$/;
    $uri .= sprintf($path_template, $self->version());

    $uri .= '?apikey=' . $self->api_key();

    foreach my $key (keys %parameters) {
        $uri .= '&' . $key . '=' . $parameters{$key};
    }

    my $ua = LWP::UserAgent->new;

    my $req = HTTP::Request->new($mode => $uri);

    my $res = $ua->request($req);

    if ($res->is_success) {
        return decode_json($res->content);
    }
    else {
        Carp::croak("Error: " . $res->status_line);
    }
}

=head1 AUTHOR

Erik Tank, C<< <erik.tank at ticketmaster.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-ticketmaster-api at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Ticketmaster-API>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Ticketmaster::API


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Ticketmaster-API>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Ticketmaster-API>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Ticketmaster-API>

=item * Search CPAN

L<http://search.cpan.org/dist/Ticketmaster-API/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2016 Erik Tank.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


=cut

1; # End of Ticketmaster::API
